### PR TITLE
Add TestFluidObjectInternal that avoids using SharedMap

### PR DESCRIPTION
**Pull request by CraigMacomber**
## Description

This change avoids impacting the alpha+legacy APIs around test objects, and just slips in a way to use an internal one which does not use SharedMap. This is then used in tree tests so they don't create needless shared maps.

TestFluidObject and ITestFluidObject are the only use @fluidframework/map in test-utils, so if it gets deprecated and removed, and replaced by the new TestFluidObjectInternal this change adds, the dependency on  @fluidframework/map could be removed, which may be enough to allow test-utils to be used to test map like it is for other DDSes.

## Reviewer Guidance

The review process is outlined on [this wiki page](https://github.com/microsoft/FluidFramework/wiki/PR-Guidelines#guidelines).

### CODE REVIEW

**Comment by anthony-murphy** [comment]
something i would call out is that this usage basically circumvents handles and instead expands the usages of name channels.  In general, i don't think we want named channels, and we really should move to a pattern that matches the aliasing pattern we use for datastores. When we don't use handles it makes gc impossible, not that we gc channels today, but someday we could. i also understand this is just for test, but unnatural test patterns tend to have a negative impact on how we build features, or expect customer to use them, as a lot of understanding of how our product works is discovered when writing tests. 

that said, i'm not totally against this change, but if there is any way to make it more obvious it is doing the wrong thing just for test i think that could be useful.

**Comment by CraigMacomber** [comment]
> something i would call out is that this usage basically circumvents handles and instead expands the usages of name channels. In general, i don't think we want named channels, and we really should move to a pattern that matches the aliasing pattern we use for datastores. When we don't use handles it makes gc impossible, not that we gc channels today, but someday we could. i also understand this is just for test, but unnatural test patterns tend to have a negative impact on how we build features, or expect customer to use them, as a lot of understanding of how our product works is discovered when writing tests.
>
> that said, i'm not totally against this change, but if there is any way to make it more obvious it is doing the wrong thing just for test i think that could be useful.

This replaces use of a named "root" channel and a shared map with a just named channels.

Is the named root channel which held the map this removes somehow better than other named channels?

Every single DataStore I have looked at uses at least one named channel, and I'm not aware of any alternative, but maybe I missed something?

**Comment by anthony-murphy** [comment]
> > something i would call out is that this usage basically circumvents handles and instead expands the usages of name channels. In general, i don't think we want named channels, and we really should move to a pattern that matches the aliasing pattern we use for datastores. When we don't use handles it makes gc impossible, not that we gc channels today, but someday we could. i also understand this is just for test, but unnatural test patterns tend to have a negative impact on how we build features, or expect customer to use them, as a lot of understanding of how our product works is discovered when writing tests.
> > that said, i'm not totally against this change, but if there is any way to make it more obvious it is doing the wrong thing just for test i think that could be useful.
> 
> This replaces use of a named "root" channel and a shared map with a just named channels.
> 
> Is the named root channel which held the map this removes somehow better than other named channels?
> 
> Every single DataStore I have looked at uses at least one named channel, and I'm not aware of any alternative, but maybe I missed something?

you are correct. they all have one, the root, or default, and then all other channels should be stored as handles within the root.

**Comment by CraigMacomber** [comment]
> you are correct. they all have one, the root, or default, and then all other channels should be stored as handles within the root.

What does "should" mean in this context. Is it like "SHOULD" as defined by https://datatracker.ietf.org/doc/html/rfc2119 ? If so what are the "full implications must" that to "be understood and carefully weighed before choosing a different course"?

Does anything break if you don't do that? Is there a reason that is a policy? Is it documented somewhere?

If GC can find the root under what ever name it has, I don't see a reason it can't find all the roots under what ever names they have.

I believe that named channels don't support reassignment, removal, synchronous creation or strong typing, but those all seem like perfectly reasonable limitations here. Is there some other concern I'm not considering, or perhaps you are just warning me about those limits which are what lots of named channels is generally not a good approach?

If I understand correctly, the old code put all these shared objects in the same datastore, so they couldn't be GCed anyway so even if they were removed from the map, wouldn't they persist as part of that datastore forever anyway? Maybe I'm missing something there?

**Comment by CraigMacomber** [comment]
The named channels created here are used at the same time the root channel would be created before this change, so upcoming plans to avoid using named channels as specific times (before attach) should not be impacted by this change.

File: packages/test/test-utils/src/testFluidObjectInternal.ts [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+	 * @param runtime - The data store runtime.
+	 * @param context - The data store context.
+	 * @param factoryEntries - A list of id to IChannelFactory mapping. For each item in the list,
+	 * a shared object is created which can be retrieved by calling getSharedObject() with the id;
+	 */
+	constructor(
+		public readonly runtime: IFluidDataStoreRuntime,
+		public readonly channel: IFluidDataStoreChannel,
+		public readonly context: IFluidDataStoreContext,
+		private readonly factoryEntriesMap: Map<string, IChannelFactory<ISharedObject>>,
```
Comment (agarwal-navin) [inline packages/test/test-utils/src/testFluidObjectInternal.ts:52]: I would suggest calling this something that indicates these are the initial set of DDSes that will be created during the data store creation and initialization process only. Maybe `initialSharedObjects` or `rootSharedObjects`. The latter also indicates that these are rooted and will not be GC'ed.

The way I see usages is this tests would create a data store with an initial set of DDSes that are rooted. It may create additional DDSes after this point but those should not be rooted (not enforced today of course) and their handle should be stored in one of the root DDSes.
Unfortunately, I don't think there are documents that explain this usage pattern. We should consider adding them. The same pattern is recommended for data stores as well.

File: packages/test/test-utils/src/testFluidObjectInternal.ts [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+	) {
+		this.handle = new FluidObjectHandle(this, "", runtime.objectsRoutingContext);
+	}
+
+	/**
+	 * Retrieves a shared object with the given id.
+	 * @param id - The id of the shared object to retrieve.
+	 */
+	public async getSharedObject(id: string): Promise<IChannel> {
```
Comment (agarwal-navin) [inline packages/test/test-utils/src/testFluidObjectInternal.ts:61]: Maybe rename this to something like `getInitialSharedObject` if it's not too much pain to update all current usages.

File: packages/test/test-utils/src/testFluidObject.ts [hunk 3]
```diff
@@ -163,6 +170,15 @@ export class TestFluidObjectFactory implements IFluidDataStoreFactory {
 	constructor(
 		private readonly factoryEntries: ChannelFactoryRegistry,
 		public readonly type = "TestFluidObjectFactory",
+		private readonly dataObjectKind: new (
```
Comment (agarwal-navin) [inline packages/test/test-utils/src/testFluidObject.ts:173]: This is temporary until `TestFluidObject` is removed right? I find this pattern a little complicated to parse.

File: packages/test/test-utils/src/testFluidObjectInternal.ts (hunk 2 above)
Comment (agarwal-navin) [reply packages/test/test-utils/src/testFluidObjectInternal.ts:61]: While looking at the usage of this in `tree/src/test/utils.ts`, I got confused for a minute as to where is the creation for this tree it's getting. Then I realized that it was created as part of initialization. Renaming this to reflect that would hopefully clear that confusion also.

File: packages/test/test-utils/src/testFluidObject.ts (hunk 3 above)
Comment (CraigMacomber) [reply packages/test/test-utils/src/testFluidObject.ts:173]: Possibly. Or we could keep it if there are other use cases for replacing what data object TestFluidObject uses.

I have factored out and documented this type for clarity.

File: packages/test/test-utils/src/testFluidObjectInternal.ts [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+	 * @param factoryEntries - A list of id to IChannelFactory mapping. For each item in the list,
+	 * a shared object is created which can be retrieved by calling {@link TestFluidObjectInternal.getInitialSharedObject} with the id;
+	 * @param initialSharedObjectsFactories - A collection of ids (which can be passed to {@link TestFluidObjectInternal.getInitialSharedObject})
+	 * and the corresponding factories to use to create the shared objects during initialization.
+	 */
+	constructor(
+		public readonly runtime: IFluidDataStoreRuntime,
+		public readonly channel: IFluidDataStoreChannel,
+		public readonly context: IFluidDataStoreContext,
+		private readonly initialSharedObjectsFactories: ReadonlyMap<
```
Comment (agarwal-navin) [inline packages/test/test-utils/src/testFluidObjectInternal.ts:55]: Rename similar property in `TestFluidObjectFactory` and `TestFluidObject` too.

**Review by agarwal-navin — APPROVED** [review]

File: packages/test/test-utils/src/testFluidObjectInternal.ts (hunk 4 above)
Comment (CraigMacomber) [reply packages/test/test-utils/src/testFluidObjectInternal.ts:55]: I guess those are not part of the stable API, so such a rename is possible. I'll do it.

**Review by agarwal-navin — DISMISSED** [review]

File: packages/test/test-utils/src/testFluidObjectInternal.ts (hunk 1 above)
Comment (agarwal-navin) [reply packages/test/test-utils/src/testFluidObjectInternal.ts]: https://dev.azure.com/fluidframework/internal/_workitems/edit/34477

**Review by agarwal-navin — APPROVED** [review]
